### DISCUSSION TITLE
V-form, Analysis and Overhaul Proposal

### DISCUSSION THREAD

**Discussion by AustinPhillipTaylor** [directus/directus#13627] [Feature Requests]
## Issue / Preface
The `v-form` component has grown to a point where the limitations of the implementation can not keep up with the demands of reasonable and expected usage. 

As Directus grows, more complex forms will become both more common and necessary. As a conerstone component of the interface, `v-form` optimization and future-proofing are considerations which should be proactively addressed in order to avoid widespread performance issues, as well as frequent and costly maintenance of an unsustainable system. 

Additionally, yet less severely, the current structure of the `v-form` component tends to incidentally restrict the form data that can be accessed in group sets, as well as the data which can be indirectly modified. This limits interface extension developers in their ability to create solutions to problems which *should* be trivial. For instance, if an individual had the requirement of updating several other fields when another was changed, it would not be possible. This shortcoming is further exacerbated by field groups, where the available field metadata is limited to only the fields present in the group. 

## Source / Cause of Issues
The source of slowdown is best understood by breaking down the flow of data. 
Below is a simplified diagram detailing the current flow of data from `v-form` instantiation down to the interface level:

<img width="2819" alt="v-form flow" src="https://user-images.githubusercontent.com/47870619/170811600-d9737ddd-0676-47b3-a99a-ed61340bfb66.png">

As you can see in the diagram, the reactive variable `formFields`, which the `v-form` template loops over in order to render the fields, is preceded by a long chain of computed dependencies. Fortunately, the `form-field` component is passed a key corresponding to the field ID, which allows Vue to intelligently avoid DOM re-renders (on a key-by-key basis if it has not changed), despite updates to `formFields`. However, the computational overhead of recalculating each computed variable dependency, up to and including `formFields` (especially when it is wasted computation because there will not be any side-effects),  is significant. 

**Further Breakdown**
The most important value to track when trying to understand the resource use implications is the `model-value` prop of the top-most `v-form` instance. 

`initial-values` is also highly involved, though it is rarely updated, and as such not much of a concern with regard to the speed of the working interface (it has more of an impact on time-to-mount, though typically negligible as well as unavoidable). 

`model-value`, on the other hand, is updated very frequently. Especially when the most-used component in forms, `v-input`, emits on `input` (i.e. every keystroke). 

With that said, during most interface updates the `values` ref (labeled 'Current values' in the diagram) will recalculate, then `fieldsParsed`, `fieldsInGroup`, and `formFields`, at which time Vue will re-loop over `formFields` to re-render any changed fields. Note  that even when there will be no DOM updates, when `formFields` updates Vue still has to loop over the variable and render everything to the shadow DOM, then check against the existing DOM, before determining what should actually be updated on the page. 

Vue is smart enough to check the calculated `formFields` value against the cached value, before trying to re-render, but we can generally ignore that optimization since, again, updates are propagated on as little as a single character input. This is a major limitation of storing the edits as a single object ref, as opposed to a dictionary of ref properties. Any change causes a comparison to fail, then the system is left trying to recalculate for every single field, when only a single field (or no field, see **Computed is lazier than you think** below) is updated. 

Furthermore, many of these dependencies use functions that iterate over some or the entire list of fields.
- `values`: Uses `Object.assign`
- `fieldsParsed`: Uses `Object.assign`, `fields.value.map` (which internally also uses a `cloneDeep` of each field, effectively another iteration over the entire field list)
- `fieldsInGroup`: Uses `fieldsParsed.value.filter`
- `formFields`: Uses `cloneDeep` on the field list, then uses `filter` on that clone, as well as an `orderBy`, and `map`. Further, before `formFields` returns it also calls a translation utility on the field list, which utilizes another `cloneDeep`, and then a `forEach` on the entries of that clone.

`assign`, `map`, `filter`, and `forEach` are all very fast on their own, however, they all still take time and effectively have to read and write for every value passed to them. This means that, at the bare minimum, for every single `model-value` update (likely a single keystroke), assuming the value has changed and will result in a DOM update, the system has to complete at least 12 iterations of the fields/values *before* Vue can iterate over the resulting field list again, to determine what needs to be updated in the DOM. 

Hopefully this adequately demonstrates how the computational resources and/or time can exponentially increase with each new field added to the list. However, it's pertinent to note that this analysis doesn't touch on the data flow in the opposite direction (i.e. the emit chain and field/data iterations therein), which further increases computational load, and subsequent slowdown. 

**Computed is lazier than you think**
One of the key benefits of using Vue's computed variables, apart from the obvious recomputation when dependencies change, is their ability to cache values and avoid pushing updates when the return value is the same. There are two main caveats here:
- When checking for value equivalency, Vue uses a vanilla strict equals check. 
- As a result of the previous point, when dependencies are objects there are often situations where the content of the dependency has not changed, but it will still trigger a recomputation.

I'll elaborate:

Imagine you have a component setup like the following:

**Template**
```html
<div>
  <h1>computedObj2: {{ computedObj2 }}</h1>
  <button @click="changeObj">Change Object</button>
</div>
```

**Setup**
```javascript
const obj = ref({ test: 'test'  });

const computedObj = computed(() =>  {
  console.log("computedObj");
  return obj.value;
});

const computedObj2 = computed(() =>  {
  console.log("computedObj2");
  return computedObj.value;
});

function changeObj() {
  console.log("changeObj");
  obj.value = Object.assign({}, obj.value);
}
```

What do you think will happen when you click the `Change Object` button, what is logged to the console?

Every single time the button is pressed, you'd log `changeObj`, then `computedObj2`, then `computedObj`. 

It may seem obvious in this simplified example, however, based on my observed usage of computed variables, it seems most people would believe you'd only log `changeObj` because `obj.value` is an equivalent object. In other words:
- Button clicked
	- Console log `changeObj` and set `obj.value` to the same value
- Template updates on next tick, accessing `computedObj2`
	- `computedObj2` is dependent on `computedObj`, so check `computedObj` before callback.
	- `computedObj` is dependent on `obj`, so check `obj` before callback.
	- `obj.value` hasn't changed, so `computedObj` won't change. 
	- `computedObj` won't change, so `computedObj2` will access cached value.

However, the real sequence of events is as follows:
- Button clicked
	- Console log `changeObj` and set `obj.value` to *new object* with the same value
- Template updates on next tick, accessing `computedObj2`
	- `computedObj2` is dependent on `computedObj`, so check `computedObj` before callback.
	- `computedObj` is dependent on `obj`, so check `obj` before callback.
	- `obj.value` has changed, so `computedObj` will recompute. 
	- `computedObj` will recompute, so `computedObj2` will recompute.
	- Run `computedObj2` callback
		- Log `computedObj2` then return `computedObj.value`
			- Run `computedObj` callback
				- Log `computedObj` then return `obj.value`

The first sequence above would be valid if the old and new value of `obj.value` could pass a strict equals check, for instance, if it was a string. However, in vanilla Javascript `{} === {}` returns false. Object instances are unique, and creating a copy of an object is akin to creating two separate symbols.

With this said, there are really only 2 situations where caching and equivalency check optimizations can be leveraged in computed variables:
- When the return value is a primitive `string`, `number`, `null`, `undefined`, or `boolean`
- Or when a manual check is performed within the computed variable. For instance, if you needed a deep comparison of an object; within the computed callback you could store a *reference* to the original object, then make a copy of the original, apply your logic, then manually compare your output to the stored reference. If equal then you return your stored reference, not your copied object, allowing it to pass a strict equals check. 

Using refs in this manner (i.e. storing objects) is only really viable for data sets with few properties, or in situations where the majority of the object's values will change at once (where bulk recalculations are integral to the component). 

In the case of Directus' `v-form`, values are inherently changed on a field-by-field basis. There's no need to update the entire computed object for a single value change.

> Earlier I mentioned that the computed variables recalculation could be triggered when no field is updated at all. This strict equals check is why. Imagine, for instance, a user copies the value of an input, then (with everything selected) pasted the same value. The field has not changed at all, but an input emit is triggered. The 'new' value triggers a series of events where, in short, object clones are created and computed values are forced to update, despite no real modifications. 

## Goals & Limitations of an Overhaul
As mentioned previously, `v-form` is a cornerstone of the Directus interface. To avoid unnecessary breaking changes, limitations must be put in place. An overhaul must:
- Retain the usage implementation of `v-form`
	- Utilize the same instantiation
	- Accept the same props
	- Maintain the same expected I/O
- Allow interfaces to emit and receive information in the same manner

In short, only the logic internal to `v-form` should be modified.

Furthermore, solutions should ideally increase stability, while reducing the data restrictions throughout the form. This could be achieved by:
- Maintaining a single source of truth
- Allowing interfaces direct access to the source-of-truth
- Maintain field values as a list of refs (This will allow updates to be synchronously handled by Vue proxies, avoiding race conditions and overwriting that occur in the current implementation when trying to emit multiple values in the same tick).

## Proposed Solution & Benefits
In order to better future-proof the `v-form` component, I propose we effectively flatten the data flow and storage by utilizing a Pinia store for saving and retrieval. 

**Drawbacks**
Saving form data in this manner may require more up-front data/field metadata parsing. In theory, this could slightly increase the initialization time for short/simple forms, though not to a significant degree. On the other hand, complex forms should universally load much more quickly (again, this is just theoretical). I believe this to be the case because there will be a requirement to do more data handling on form initialization, however, the data will generally only need to be iterated over just once or twice, for the life of the form. 

**Benefits**
Using a Pinia store for form field handling comes with many benefits including:
- Simplified and predictable data structure
	- Debugging will be easier as Vue DevTools will be able to display our store values
- Separation of concerns - All field handling getters/setters can be defined in the form store, simplifying the `v-form` structure.
- Improve code reusability
- Improve field flexibility
	- An interface could update multiple fields at once without worrying about data flow
	- Opens the path for more complex situations, for instance, dynamic groups. I can't think of a particular use-case for this at the moment, but being able to change field groups on-the-fly, in a live form, could allow for more creative interfaces (the methodology for doing so should be clear by observing how groups will be handled, in the diagram under **Implementation**). 
	- In conjunction with the previous point, there will be less restriction on how data can be handled, without any extra work. Uses, such as dynamic groups, are simply a side-effect of the manner in which fields will be handled. 
- It goes without saying that this approach would be much more performant. If for no other reason, storing field values as individual refs will allow leveraging primitive value caching on all applicable fields.
- Probably the second-most important benefit, simplifying the structure in this manner will make it much easier to understand the flow of data, from a developer's standpoint. In other words, bugs will be more easily caught before deployment, and less time will be spent on patching small edge-cases. 

**Implementation**

Below is a simplified diagram demonstrating how the form data will flow with this shift to a Pinia store. 

<img width="4241" alt="v-form proposal" src="https://user-images.githubusercontent.com/47870619/171066249-55e6860d-de25-499a-9853-183f8d634d04.png">

At first glance, the structure appears to be much more complex than the current flow of the `v-form` component. This is largely due to much of the complexity of the current flow being condensed into the various `composables`. 

As well, it's important to reiterate that the two diagrams presented in this proposal both only demonstrate how data will flow down, from instantiation to form field interface. As a result, the full scale of the current implementation is not immediately apparent. Because we currently use a trickle-down approach for the form data (i.e. passing down and parsing props to each subsequent level), along with an emit chain (i.e. emitting and parsing events and data up each level of the component tree when a change is made), the diagram for the current model would be better represented by multiple layers (almost recursive) of the same data flow. 

Conversely, the diagram for the proposed form flow is mostly inclusive of the entire data flow (discrepancies are detailed below). As well, with a single source-of-truth store, data fetching can be delegated to the lowest common denominator of all fields (the `form-field` component). As a result, the upward data movement (the emit chain) is almost completely removed from the equation.

**Details of the Proposed Flow**

Step-by-step:
- `v-form` is initialized
	- If form instance is first in chain (we could determine this by checking if the store is empty and maybe setting an extra `isPopulating` value) we'll attach a callback to reset the entire form store when the form instance is unmounted.
	- Form store is retrieved and populate action is called with props.
		- If `fields`/`initial values`/`model values` are passed, loop over them. If field IDs are already populated in `fields` state (such as in the case of a nested form), ignore them (i.e. do not try repopulating). Within all field objects, set all metadata necessary for field usage (i.e. editable, hidden, initial value, default value, etc). 
		- While looping over fields, if there is a `group` set, and there exists a group field with a corresponding ID, create/append the applicable `group` state item. 
		- When finished populating, set `fieldsPopulated` state ref value to true.
	- Initialize form template with a conditional `v-if="fieldsPopulated"`
	- Within template, loop over fields from form store.
		- If the `group` prop is set on `v-form` instance, retrieve group ID array from form store and loop over matching fields.
		- If field is part of a group, and the `group` prop is not set, ignore the field.
		- If the field *is* a group, render group interface, passing down group ID. 
		- If the field is *not* part of a group, and the `group` prop is not set, move on to `form-field` component (All fields should reach this point unless the group interface does not instantiate a nested `v-form` instance).
		- **form-field** (receives field ID in props)
			- Fetches field value ref from `fields` state (note that we store a reference to the value, we don't copy it or encapsulate it in another computed function). 
			- Field value is passed down as prop to field interface
				- Field interfaces will act exactly as the do in the current form flow, receiving the field value and emitting update events. The parent `form-field` component handles all of the reactivity with the form store. 
			- When the field interface emits an update, `form-field` will directly update the ref in the form store, via the stored reference (this means that the field value will automatically be propagated by Vue to all interfaces that access the ref).

To put it simply, that's it. No other data is emitted upward, and unless the initial props are modified (apart from the `model-value`), the entire list of fields will never be looped over again, so long as the form is mounted. 

> **Note:** There should be a discussion regarding how groups should be handled. Internally, all groups already pass down a `group` prop to their nested `v-form` instance. Apart from minor variable name changes, nothing should break there, in fact we can remove a lot of the redundant `fields`, `initial-values`, etc. props passed down to these nested forms.
. 
However, despite not breaking any groups internally, this could be a breaking change for 3rd-party group interfaces. The current setup passes down all field data to group interfaces via props. Opting to only pass down the group ID would break the interfaces that make use of that information. 
.
Fortunately, this approach will allow global access to the form store and all of its data (which we can make even simpler by `providing` the `fields` and `groups` state items). So the avenue to access all of the necessary data will still be there. 
.
Suffice to say, a decision will need to be made regarding adding an extra layer before the group interface (similar to the relationship between `form-field` and the form interface) in order to resolve and pass down the field data in props. Or if it would just be better to keep the flow simplified and deprecate that method of passing data down to group interfaces. (To put it more simply, should the `group` prop be the only prop passed to group interfaces?)

---

One small detail you may have noticed - the field value ref is outlined in orange, labeled a "Re-Proxied Ref." This is important for avoiding a mountain of callbacks, and delegating all field data handling to the form store. I'll elaborate:

Firstly, a brief refresher, all Vue refs work by creating proxies. They intercept the `get` and `set` functionality of an object property, in this case specifically getting or setting the `value` property, in order to filter the data and/or run callbacks after the data has changed. What's good to know is that any object-like value can be proxied in Javascript, and proxies themselves fall under that umbrella. 

With this said, my idea here is when the field values are populated we'll create the obvious Vue ref. However, the actual value set to the field value property will be a proxy of that ref. 

> To avoid confusion since both Vue refs and our field use a `value` property, from here on out I'll refer to the `value` property within the field data as `f_value`, and the `value` property within the Vue ref as `v_value`. 

What this means is that before Vue does its thing we can intercept/filter the action any time there is an attempt to set `v_value`. Once that is done, we can utilize the native `Reflect` class in order to continue passing the data on to the Vue proxy handler ([More information on Proxies and the Reflect class can be found here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy)). 

In doing this, we can more easily modify the `changes` state value concurrently with the `v_value` updates. The flow would be like the following:
- Field interface emits update
- `form-field` component runs `f_value.value = emitted_value`
- If the field value is not changeable, it will stop here.
- ... Do any number of other preliminary checks and if they don't pass then it will stop here.
- Our custom proxy will check the new value against the default value, initial value, and/or the value already set in `changes`. It will proceed to set, update, or remove the entry from `changes` as applicable. 
- If everything passed successfully, `Reflect` the Vue proxy which will update and propagate the ref as it otherwise would have. 

## Conclusion / Summation

There are a few minor things I've left out, but I hope this gives you the general idea of how a system like this could work, and the benefits it would bring. 

The majority of the proposed changes are non-breaking, and those that aren't can either be accounted for with a little extra code, or deprecated entirely since they're minor, and new methods will be in place to achieve the same outcome in a better, more standardized manner. 

As you can imagine, it's taken quite a long time to write up this proposal, so I'm sure there are plenty of areas I could've explained better, or instances where I flat-out forgot details. So please feel free to ask about anything, be it clarification, suggestions, or critique. I've taken so much time to write this up because I truly believe this is an important portion of Directus that needs attention, and I want to ensure that the discussion around it is thorough, even if we entirely decide against this approach.

Thank you for your consideration!

**Comment by br41nslug** [comment]
The analysis in this proposal was invaluable for getting my short term performance fix in place ❤️ Would love to see if we can plan some of these proposed changes as long term improvements to reduce complexity in the beast we call `v-form`

**Comment by jacobrienstra** [comment]
This is amazing, thank you so much for writing this up! I think this might make calculated fields/values nearly trivial as well. Fully support this approach.

I'm curious, how do you imagine handling nested forms in drawer-items, for relational interfaces. Would they also use a "`group-id`" to group the relevant fields, even though it's not a "field-group" per se, or would there be another way to have multiple form stores tracked?

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
